File: frontend/lib/services/weight_service.dart
import '../models/weight_record.dart';
import 'api_service.dart';

class WeightService {
  final ApiService _apiService = ApiService();

  /// 创建体重记录
  Future<WeightRecord> createRecord({
    required int userId,
    required double weight,
    double bodyFat = 0,
    double muscle = 0,
    double water = 0,
    double bmi = 0,
    String note = '',
    DateTime? measuredAt,
  }) async {
    final response = await _apiService.post('/weight/record', {
      'user_id': userId,
      'weight': weight,
      'body_fat': bodyFat,
      'muscle': muscle,
      'water': water,
      'bmi': bmi,
      'note': note,
      'measured_at': (measuredAt ?? DateTime.now()).toUtc().toIso8601String(),
    });

    if (response.statusCode == 201) {
      return WeightRecord.fromJson(response.data);
    } else {
      throw Exception('创建体重记录失败');
    }
  }

  /// 获取体重记录列表
  Future<List<WeightRecord>> getRecords({
    required int userId,
    DateTime? startDate,
    DateTime? endDate,
  }) async {
    final params = <String, String>{
      'user_id': userId.toString(),
    };

    if (startDate != null) {
      params['start_date'] = startDate.toIso8601String().split('T').first;
    }
    if (endDate != null) {
      params['end_date'] = endDate.toIso8601String().split('T').first;
    }

    final response = await _apiService.get('/weight/records', queryParameters: params);

    if (response.statusCode == 200) {
      final data = response.data;
      List recordsList = data['records'] ?? [];
      return recordsList.map((r) => WeightRecord.fromJson(r)).toList();
    } else {
      throw Exception('获取体重记录失败');
    }
  }

  /// 获取体重趋势
  Future<WeightTrend> getTrend({
    required int userId,
    int days = 30,
  }) async {
    final response = await _apiService.get(
      '/weight/trend',
      queryParameters: {
        'user_id': userId.toString(),
        'days': days.toString(),
      },
    );

    if (response.statusCode == 200) {
      return WeightTrend.fromJson(response.data);
    } else {
      throw Exception('获取体重趋势失败');
    }
  }

  /// 更新体重记录
  Future<WeightRecord> updateRecord({
    required int id,
    double? weight,
    double? bodyFat,
    double? muscle,
    double? water,
    double? bmi,
    String? note,
    DateTime? measuredAt,
  }) async {
    final body = <String, dynamic>{};

    if (weight != null) body['weight'] = weight;
    if (bodyFat != null) body['body_fat'] = bodyFat;
    if (muscle != null) body['muscle'] = muscle;
    if (water != null) body['water'] = water;
    if (bmi != null) body['bmi'] = bmi;
    if (note != null) body['note'] = note;
    if (measuredAt != null) body['measured_at'] = measuredAt.toUtc().toIso8601String();

    final response = await _apiService.put('/weight/record/$id', body);

    if (response.statusCode == 200) {
      return WeightRecord.fromJson(response.data);
    } else {
      throw Exception('更新体重记录失败');
    }
  }

  /// 删除体重记录
  Future<void> deleteRecord(int id) async {
    final response = await _apiService.delete('/weight/record/$id');

    if (response.statusCode != 200) {
      throw Exception('删除体重记录失败');
    }
  }
}

File: backend/internal/services/weight_service.go
package services

import (
	"time"

	"github.com/your-org/loss-weight/backend/internal/models"
	"go.uber.org/zap"
	"gorm.io/gorm"
)

type WeightService struct {
	db     *gorm.DB
	logger *zap.Logger
}

func NewWeightService(db *gorm.DB, logger *zap.Logger) *WeightService {
	return &WeightService{
		db:     db,
		logger: logger,
	}
}

type CreateWeightRecordRequest struct {
	UserID     uint      `json:"user_id" binding:"required"`
	Weight     float32   `json:"weight" binding:"required"`
	BodyFat    float32   `json:"body_fat"`
	Muscle     float32   `json:"muscle"`
	Water      float32   `json:"water"`
	BMI        float32   `json:"bmi"`
	Note       string    `json:"note"`
	MeasuredAt time.Time `json:"measured_at"`
}

type UpdateWeightRecordRequest struct {
	Weight     float32   `json:"weight"`
	BodyFat    float32   `json:"body_fat"`
	Muscle     float32   `json:"muscle"`
	Water      float32   `json:"water"`
	BMI        float32   `json:"bmi"`
	Note       string    `json:"note"`
	MeasuredAt time.Time `json:"measured_at"`
}

func (s *WeightService) CreateRecord(req *CreateWeightRecordRequest) (*models.WeightRecord, error) {
	record := &models.WeightRecord{
		UserID:     req.UserID,
		Weight:     req.Weight,
		BodyFat:    req.BodyFat,
		Muscle:     req.Muscle,
		Water:      req.Water,
		BMI:        req.BMI,
		Note:       req.Note,
		MeasuredAt: req.MeasuredAt,
	}

	if req.MeasuredAt.IsZero() {
		record.MeasuredAt = time.Now()
	}

	if err := s.db.Create(record).Error; err != nil {
		s.logger.Error("failed to create weight record", zap.Error(err))
		return nil, err
	}

	return record, nil
}

func (s *WeightService) GetRecordsByUser(userID uint, startDate, endDate time.Time) ([]models.WeightRecord, error) {
	var records []models.WeightRecord
	query := s.db.Where("user_id = ?", userID)

	if !startDate.IsZero() {
		query = query.Where("measured_at >= ?", startDate)
	}
	if !endDate.IsZero() {
		// Half-open: endDate is exclusive. Handler converts the inclusive
		// client-supplied YYYY-MM-DD into next-day midnight (in client tz).
		query = query.Where("measured_at < ?", endDate)
	}

	if err := query.Order("measured_at DESC").Find(&records).Error; err != nil {
		s.logger.Error("failed to get weight records", zap.Error(err))
		return nil, err
	}

	return records, nil
}

func (s *WeightService) GetRecordByID(id uint) (*models.WeightRecord, error) {
	var record models.WeightRecord
	if err := s.db.First(&record, id).Error; err != nil {
		s.logger.Error("failed to get weight record", zap.Error(err))
		return nil, err
	}
	return &record, nil
}

func (s *WeightService) UpdateRecord(id uint, req *UpdateWeightRecordRequest) (*models.WeightRecord, error) {
	var record models.WeightRecord
	if err := s.db.First(&record, id).Error; err != nil {
		s.logger.Error("failed to get weight record", zap.Error(err))
		return nil, err
	}

	updates := make(map[string]interface{})
	if req.Weight > 0 {
		updates["weight"] = req.Weight
	}
	if req.BodyFat > 0 {
		updates["body_fat"] = req.BodyFat
	}
	if req.Muscle > 0 {
		updates["muscle"] = req.Muscle
	}
	if req.Water > 0 {
		updates["water"] = req.Water
	}
	if req.BMI > 0 {
		updates["bmi"] = req.BMI
	}
	if req.Note != "" {
		updates["note"] = req.Note
	}
	if !req.MeasuredAt.IsZero() {
		updates["measured_at"] = req.MeasuredAt
	}

	if err := s.db.Model(&record).Updates(updates).Error; err != nil {
		s.logger.Error("failed to update weight record", zap.Error(err))
		return nil, err
	}

	return &record, nil
}

func (s *WeightService) DeleteRecord(id uint) error {
	if err := s.db.Delete(&models.WeightRecord{}, id).Error; err != nil {
		s.logger.Error("failed to delete weight record", zap.Error(err))
		return err
	}
	return nil
}

func (s *WeightService) GetWeightTrend(userID uint, days int) ([]models.WeightRecord, error) {
	startDate := time.Now().AddDate(0, 0, -days)
	var records []models.WeightRecord

	if err := s.db.Where("user_id = ? AND measured_at >= ?", userID, startDate).
		Order("measured_at ASC").
		Find(&records).Error; err != nil {
		s.logger.Error("failed to get weight trend", zap.Error(err))
		return nil, err
	}

	return records, nil
}
